### ISSUE TITLE
playing sine wave on raspberry pi has 25+ seconds delay before sound starts to play

### ISSUE THREAD

**Issue by nksan** [spatialaudio/python-sounddevice#327]
I have been testing sounddevice to play a sine wave (based on  the code example in the docs).
Everything works - but after I start the output stream, it takes about 25-30 seconds before the sine wave sound starts to play.
On a Raspberry pi 3B - I did a clean latest rasbian buster light install - running headless (SSH from mac), and only installed what was necessary for sounddevice and soundfile. Installed python 3.9.4 - but I had the same problem when I was running 3.7 that came with the OS (playing the wave file example).  I can use aplay on the RPi to play a sound wave no problem (no delay)

I initially used the code from sine wave example `with outputstream(...)` - and but then changed to creating the stream first and start it separately: same result.
I inserted a bunch of calls to time.time() - I can see the stream starts right away but the first call to the callback takes 25-30 seconds.

My Code:

```
import sounddevice as sd
import numpy as np 

start_idx = 0
cnt=1
sd.default.device=0


try:
    samplerate = 44100

    def callback(outdata, frames, time, status):
        if status:
            print(status, file=sys.stderr)
        global start_idx
        global cnt
        t = (start_idx + np.arange(frames)) / samplerate
        t = t.reshape(-1, 1)
        outdata[:] = 1 * np.sin(2 * np.pi * 440 * t) 
        start_idx += frames
        cnt+=1
        if cnt>20: 
            print('count =',cnt)
            raise sd.CallbackStop

    myStream=sd.OutputStream(device=0, channels=1, callback=callback,
                         samplerate=samplerate)
    myStream.start()
    print('stream is active: ',myStream.active).   #This shows True
    print('cpu load = ',myStream.cpu_load).        #This shows 0.0
    input()  # (I wait until I see count=21 printed in terminal and sound stops in headphones, then hit return )
    myStream.stop()
    myStream.close()
```
For info - sd.query_devices() yields:
```
* 0 bcm2835 Headphones: - (hw:0,0), ALSA (0 in, 8 out)
  1 Logitech USB Microphone: Audio (hw:1,0), ALSA (1 in, 0 out)
  2 sysdefault, ALSA (0 in, 128 out)
  3 output, ALSA (0 in, 8 out)
  4 dmix, ALSA (0 in, 2 out)
  5 default, ALSA (0 in, 128 out)
```

any ideas why this happens? could it be a PortAudio issue on the pi - nothing to do with sounddevice?

**Comment by HaHeho** [comment]
That code was not executable, so I don't know how it ever ran (even after a couple of seconds). Try this
```python
import sys

import sounddevice as sd
import numpy as np

start_idx = 0
cnt = 1
sd.default.device = 0
samplerate = 44100


def callback(outdata, frames, time, status):
    if status:
        print(status, file=sys.stderr)
    global start_idx
    global cnt
    t = (start_idx + np.arange(frames)) / samplerate
    t = t.reshape(-1, 1)
    outdata[:] = .05 * np.sin(2 * np.pi * 440 * t)
    start_idx += frames
    cnt += 1
    if cnt > 20:
        print('count =', cnt)
        raise sd.CallbackStop


# print(sd.query_devices())
myStream = sd.OutputStream(device='Built-in Output', channels=1,
                           callback=callback, samplerate=samplerate)
myStream.start()
print('stream is active: ', myStream.active)  # This shows True
print('cpu load = ', myStream.cpu_load)  # This shows 0.0
input()  # (I wait until I see count=21 printed in terminal and sound stops in headphones, then hit return )
myStream.stop()
myStream.close()
```

If it still does not work, or just delayed, put some logging prints at different positions in order to see where the delay might arise.

**Comment by nksan** [comment]
HI,
 not sure why the code was not executable - although I see that I forgot  to print the last two lines  hereto close the `try block`. (but they were in my code - so the code ran).
code:
```
except Exception as e:
    pass
```
IF I understand the above - the code you posted is a copy of mine, without the try block (not really needed - you're right) - and with:
`device='Built-in Output'`
as argument of the stream.  

I ran your code exactly as you posted and I got a valueError
```
Traceback (most recent call last):
  File "/home/pi/projects/looper/sdtest.py", line 28, in <module>
    myStream = sd.OutputStream(device='Built-in Output', channels=1,
  File "/home/pi/.local/lib/python3.9/site-packages/sounddevice.py", line 1488, in __init__
    _StreamBase.__init__(self, kind='output', wrap_callback='array',
  File "/home/pi/.local/lib/python3.9/site-packages/sounddevice.py", line 811, in __init__
    _get_stream_parameters(kind, device, channels, dtype, latency,
  File "/home/pi/.local/lib/python3.9/site-packages/sounddevice.py", line 2650, in _get_stream_parameters
    device = _get_device_id(device, kind, raise_on_error=True)
  File "/home/pi/.local/lib/python3.9/site-packages/sounddevice.py", line 2790, in _get_device_id
    raise ValueError(
ValueError: No output device matching 'Built-in Output'
```
if I replace 'Built-in Output' with the number 0 - as per my original code:
`device=0`
as argument for the outputstream call,  then the code runs fine -
but I gets the same behavior that I described:  the stream is started and shows alive, but it takes 25-30 seconds before the first call to callback is made and the sound starts to play.

Before I posted - I put in lotsof print statements and timestamped everything with time.time() - and basically I was able to see that the stream was initiliazed and started correctly - but then it would take 25 to 30 seconds before the callback function was called.  After this - the callback function was called repeatedly correctly so that the sine wave would play without problems.

It's just this 25 second delay before the sound plays that has got me stumped.

**Comment by nksan** [comment]
Thanks Mathias - I had not found those. at least I know I'm not crazy...
but like you said - no solution.  Have you got any ideas of how I might approach this?

**Comment by mgeier** [comment]
The mailing list thread I've linked to above mentions that the problem also appears when not using the `sounddevice` module, so I guess this will have to be fixed in the underlying PortAudio library.

Have you tried the latest PortAudio release v19.7.0 (http://files.portaudio.com/download.html)?

There might not be a package yet, you might have to compile it from source.

If that doesn't help you can try to continue the mentioned mailing list thread or open an issue at https://github.com/PortAudio/portaudio.

**Comment by nksan** [comment]
HI,
I removed the package portaudio with apt-get  - and installed the newest one compiling from source using configure && make etc.  
checked that version is now 19.7 within sounddevice

same result: 25 seconds delay...

**Comment by mgeier** [comment]
Thanks for testing this, it's good to know.

There is nothing more for me to do here, the rest has to be done at the PortAudio project.

**Comment by nksan** [comment]
@josephernest - Sorry I did not.  It looked from the above that the issue is in the portaudio library itself - and I did not find much traction there.
In the end I stopped using sounddevice and went directly with alsaaudio, audioop, and wave modules to implement what I needed in managing samples and playing them back.  I works fine although alsa has its own quirks - but I was able to find enough articles etc. to deal with anything that came up.
